File: svx/source/table/propertyset.cxx
/* -*- Mode: C++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
/*
 * This file is part of the LibreOffice project.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * This file incorporates work covered by the following license notice:
 *
 *   Licensed to the Apache Software Foundation (ASF) under one or more
 *   contributor license agreements. See the NOTICE file distributed
 *   with this work for additional information regarding copyright
 *   ownership. The ASF licenses this file to you under the Apache
 *   License, Version 2.0 (the "License"); you may not use this file
 *   except in compliance with the License. You may obtain a copy of
 *   the License at http://www.apache.org/licenses/LICENSE-2.0 .
 */


#include "propertyset.hxx"

using namespace ::com::sun::star::uno;
using namespace ::com::sun::star::beans;
using namespace ::com::sun::star::lang;

namespace sdr::table {

FastPropertySetInfo::FastPropertySetInfo( const PropertyVector& rProps )
{
    addProperties( rProps );
}


FastPropertySetInfo::~FastPropertySetInfo()
{
}


void FastPropertySetInfo::addProperties( const PropertyVector& rProps )
{
    sal_uInt32 nIndex = maProperties.size();
    sal_uInt32 nCount = rProps.size();
    maProperties.resize( nIndex + nCount );
    for( const Property& rProperty : rProps )
    {
        maProperties[nIndex] = rProperty;
        maMap[ rProperty.Name ] = nIndex++;
    }
}


const Property& FastPropertySetInfo::getProperty( const OUString& aName )
{
    PropertyMap::iterator aIter( maMap.find( aName ) );
    if( aIter == maMap.end() )
        throw UnknownPropertyException( aName, static_cast<cppu::OWeakObject*>(this));
    return maProperties[(*aIter).second];
}


const Property* FastPropertySetInfo::hasProperty( const OUString& aName )
{
    PropertyMap::iterator aIter( maMap.find( aName ) );
    if( aIter == maMap.end() )
        return nullptr;
    else
        return &maProperties[(*aIter).second];
}


// XPropertySetInfo


Sequence< Property > SAL_CALL FastPropertySetInfo::getProperties()
{
    return Sequence< Property >( maProperties.data(), maProperties.size() );
}


Property SAL_CALL FastPropertySetInfo::getPropertyByName( const OUString& aName )
{
    return getProperty( aName );
}


sal_Bool SAL_CALL FastPropertySetInfo::hasPropertyByName( const OUString& aName )
{
    return hasProperty( aName ) != nullptr;
}

FastPropertySet::FastPropertySet( const rtl::Reference< FastPropertySetInfo >& xInfo )
: mxInfo( xInfo )
{
}


FastPropertySet::~FastPropertySet()
{
}


// XPropertySet


Reference< XPropertySetInfo > SAL_CALL FastPropertySet::getPropertySetInfo(  )
{
    return mxInfo;
}


void SAL_CALL FastPropertySet::setPropertyValue( const OUString& aPropertyName, const Any& aValue )
{
    setFastPropertyValue( mxInfo->getProperty( aPropertyName ).Handle, aValue );
}


Any SAL_CALL FastPropertySet::getPropertyValue( const OUString& aPropertyName )
{
    return getFastPropertyValue( mxInfo->getProperty( aPropertyName ).Handle );
}


void SAL_CALL FastPropertySet::addPropertyChangeListener( const OUString&, const Reference< XPropertyChangeListener >& )
{
}


void SAL_CALL FastPropertySet::removePropertyChangeListener( const OUString&, const Reference< XPropertyChangeListener >& )
{
}


void SAL_CALL FastPropertySet::addVetoableChangeListener( const OUString&, const Reference< XVetoableChangeListener >& )
{
}


void SAL_CALL FastPropertySet::removeVetoableChangeListener( const OUString&, const Reference< XVetoableChangeListener >& )
{
}


// XMultiPropertySet


void SAL_CALL FastPropertySet::setPropertyValues( const Sequence< OUString >& aPropertyNames, const Sequence< Any >& aValues )
{
    if( aPropertyNames.getLength() != aValues.getLength() )
        throw IllegalArgumentException();

    const Any* pValues = aValues.getConstArray();
    for( const OUString& rPropertyName : aPropertyNames )
    {
        const Property* pProperty = mxInfo->hasProperty( rPropertyName );
        if( pProperty ) try
        {
            setFastPropertyValue( pProperty->Handle, *pValues );
        }
        catch( UnknownPropertyException& )
        {
        }
        pValues++;
    }
}


Sequence< Any > SAL_CALL FastPropertySet::getPropertyValues( const Sequence< OUString >& aPropertyNames )
{
    sal_Int32 nCount = aPropertyNames.getLength();
    Sequence< Any > aValues( nCount );

    Any* pValues = aValues.getArray();
    for( const OUString& rPropertyName : aPropertyNames )
    {
        const Property* pProperty = mxInfo->hasProperty( rPropertyName );
        if( pProperty ) try
        {
            *pValues = getFastPropertyValue( pProperty->Handle );
        }
        catch( UnknownPropertyException& )
        {
        }
        pValues++;
    }
    return aValues;
}


void SAL_CALL FastPropertySet::addPropertiesChangeListener( const Sequence< OUString >&, const Reference< XPropertiesChangeListener >& )
{
}


void SAL_CALL FastPropertySet::removePropertiesChangeListener( const Reference< XPropertiesChangeListener >& )
{
}


void SAL_CALL FastPropertySet::firePropertiesChangeEvent( const Sequence< OUString >&, const Reference< XPropertiesChangeListener >& )
{
}

}

/* vim:set shiftwidth=4 softtabstop=4 expandtab: */

File: connectivity/java/dbtools/src/org/apache/openoffice/comp/sdbc/dbtools/comphelper/PropertySet.java
/**************************************************************
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 *************************************************************/

package com.sun.star.sdbcx.comp.postgresql.comphelper;

import com.sun.star.beans.PropertyVetoException;
import com.sun.star.beans.UnknownPropertyException;
import com.sun.star.beans.XFastPropertySet;
import com.sun.star.beans.XMultiPropertySet;
import com.sun.star.beans.XPropertiesChangeListener;
import com.sun.star.beans.XPropertyChangeListener;
import com.sun.star.beans.XPropertySet;
import com.sun.star.beans.XPropertySetInfo;
import com.sun.star.beans.XVetoableChangeListener;
import com.sun.star.lang.IllegalArgumentException;
import com.sun.star.lang.WrappedTargetException;
import com.sun.star.lib.uno.helper.ComponentBase;
import com.sun.star.sdbcx.comp.postgresql.comphelper.PropertySetAdapter.PropertyGetter;
import com.sun.star.sdbcx.comp.postgresql.comphelper.PropertySetAdapter.PropertySetter;
import com.sun.star.uno.Type;

public class PropertySet extends ComponentBase implements XPropertySet, XFastPropertySet, XMultiPropertySet {
    private final PropertySetAdapter impl;

    protected PropertySet() {
        impl = new PropertySetAdapter(this, this);
    }

    @Override
    protected void postDisposing() {
        impl.dispose();
    }

    public void registerProperty(String propertyName, int handle, Type type, short attributes, PropertyGetter getter, PropertySetter setter) {
        impl.registerProperty(propertyName, handle, type, attributes, getter, setter);
    }

    public void registerProperty(String propertyName, Type type, short attributes, PropertyGetter getter, PropertySetter setter) {
        impl.registerProperty(propertyName, type, attributes, getter, setter);
    }

    public synchronized void addPropertyChangeListener(String propertyName, XPropertyChangeListener listener)
            throws UnknownPropertyException, WrappedTargetException {
        // only add listeners if you are not disposed
        if (!isDisposed()) {
            impl.addPropertyChangeListener(propertyName, listener);
        }
    }

    public synchronized void addVetoableChangeListener(String propertyName, XVetoableChangeListener listener)
            throws UnknownPropertyException, WrappedTargetException {
        // only add listeners if you are not disposed
        if (!isDisposed()) {
            impl.addVetoableChangeListener(propertyName, listener);
        }
    }

    public synchronized void addPropertiesChangeListener(String[] propertyNames, XPropertiesChangeListener listener) {
        // only add listeners if you are not disposed
        if (!isDisposed()) {
            impl.addPropertiesChangeListener(propertyNames, listener);
        }
    }

    public XPropertySetInfo getPropertySetInfo() {
        return impl.getPropertySetInfo();
    }

    public synchronized Object getPropertyValue(String propertyName) throws UnknownPropertyException, WrappedTargetException {
        checkDisposed();
        return impl.getPropertyValue(propertyName);
    }

    @Override
    public synchronized Object getFastPropertyValue(int handle) throws UnknownPropertyException, WrappedTargetException {
        checkDisposed();
        return impl.getFastPropertyValue(handle);
    }

    public synchronized Object[] getPropertyValues(String[] propertyNames) {
        checkDisposed();
        return impl.getPropertyValues(propertyNames);
    }

    public synchronized void removePropertyChangeListener(String propertyName, XPropertyChangeListener listener)
            throws UnknownPropertyException, WrappedTargetException {
        // all listeners are automatically released in a dispose call
        if (!isDisposed()) {
            impl.removePropertyChangeListener(propertyName, listener);
        }
    }

    public synchronized void removeVetoableChangeListener(String propertyName, XVetoableChangeListener listener)
            throws UnknownPropertyException, WrappedTargetException {
        // all listeners are automatically released in a dispose call
        if (!isDisposed()) {
            impl.removeVetoableChangeListener(propertyName, listener);
        }
    }

    public synchronized void removePropertiesChangeListener(XPropertiesChangeListener listener) {
        // all listeners are automatically released in a dispose call
        if (!isDisposed()) {
            impl.removePropertiesChangeListener(listener);
        }
    }

    public synchronized void setPropertyValue(String propertyName, Object value)
            throws UnknownPropertyException, PropertyVetoException, IllegalArgumentException, WrappedTargetException {
        checkDisposed();
        impl.setPropertyValue(propertyName, value);
    }

    public synchronized void setFastPropertyValue(int handle, Object value)
            throws UnknownPropertyException, PropertyVetoException, IllegalArgumentException, WrappedTargetException {
        checkDisposed();
        impl.setFastPropertyValue(handle, value);
    }

    public synchronized void setPropertyValues(String[] propertyNames, Object[] values)
            throws PropertyVetoException, IllegalArgumentException, WrappedTargetException {
        checkDisposed();
        impl.setPropertyValues(propertyNames, values);
    }

    public synchronized void firePropertiesChangeEvent(String[] propertyNames, XPropertiesChangeListener listener) {
        checkDisposed();
        impl.firePropertiesChangeEvent(propertyNames, listener);
    }
}
